src/mappings/billing.ts

```diff
@@ -1,28 +1,42 @@
-import { Billing, User, TokensAdded, TokensRemoved, TokensPulled } from '../types/schema'
+import {
+  TokensAdded,
+  TokensRemoved,
+  TokensPulled,
+  InsufficientBalanceForRemoval,
+  Collector,
+} from '../types/schema'
+import { store } from '@graphprotocol/graph-ts'
 import {
   TokensAdded as AddedEvent,
   TokensRemoved as RemovedEvent,
   TokensPulled as PulledEvent,
-  GatewayUpdated,
+  InsufficientBalanceForRemoval as InsufficientBalanceForRemovalEvent,
+  CollectorUpdated,
   NewOwnership,
 } from '../types/Billing/Billing'
-import { BigInt, Address } from '@graphprotocol/graph-ts'
+
 import {
   getBilling,
   createOrLoadUser,
   getAndUpdateBillingDailyData,
   getAndUpdateUserDailyData,
+  DEFAULT_BILLING_ID,
 } from './helpers'
 
 /**
- * @dev handleEpochRun - Sets the gateway on the Billing Entity. Creates entity on first try
+ * @dev handleEpochRun - Sets the gateways ("collectors") on the Billing Entity. Creates entity on first try
  */
-export function handleGatewayUpdated(event: GatewayUpdated): void {
-  let billing = getBilling(event.address)
-  billing.gateway = event.params.newGateway
-
+export function handleCollectorUpdated(event: CollectorUpdated): void {
+  let collector = Collector.load(event.params.collector.toHexString())
+  if (collector == null && event.params.enabled) {
+    collector = new Collector(event.params.collector.toHexString())
+    collector.billing = DEFAULT_BILLING_ID
+    collector.save()
+  } else if (collector != null && !event.params.enabled) {
+    store.remove('Collector', collector.id)
+  }
```

Comment:
I wouldn't delete entities unless it's completely required for stuff to work properly.
This could be easily swapped to a flag inside the entity that clearly outlines if an entity is `enabled` or not, which would make this explicit, but still retain entities in case we want to query old entities for their state

Reply:
I'll let @IainM22 weigh in on whether this affects the UI in any way, but otherwise this makes sense to me so I'd work on a fix

Reply:
Pushed this fix in the meantime, and I'll check with Iain to see if it works for him

Reply:
Fix makes sense, so it should most likely work for @IainM22 :)

Reply:
Yup he confirmed so we're good to go 👍